### PR TITLE
New erx 6

### PR DESCRIPTION
**Pull request by juggernautsei**
Posting PR as requested

Up for grabs demo here:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.135

### CODE REVIEW

**Comment by bradymiller** [comment]
@juggernautsei ,
I think this is getting close :)
Is this code testing well?

**Comment by juggernautsei** [comment]
I am going to test a fresh install from scratch and see if the code works. I will update you later.

**Comment by bradymiller** [comment]
Just created an up for grabs demo for this PR here:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.135

**Comment by juggernautsei** [comment]
This is not testing well right now. The provider data is not populating. I feel like I am troubleshooting someone else's code. Trying to figure out how the doc info gets populated. Update you as soon as I get this fixed.

**Comment by juggernautsei** [comment]
The system is testing well now. 
Weno Exchange LLC|2017-08-31|01:55:49.56314Z|001|Transaction successful||

Anyone that follows the instructions for testing should receive this message.

File: interface/weno/formularyInsert.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+$zip->open('FormularyFiles.zip');
+$zip->extractTo($GLOBALS['temporary_files_dir']);
+$zip->close();
+
+ $file = $GLOBALS['temporary_files_dir'] . "./Weno Exchange LLC/BSureFormulary.txt";
+ $lines = count(file($file));
+ echo text($lines). "<br>";
+  
+ $i = 2;
+
```
Comment (bradymiller) [inline interface/weno/formularyInsert.php]: To drastically speed up this import of data, place the following lines here:
``````
// Settings to drastically speed up import with InnoDB
sqlStatementNoLog("SET autocommit=0");
sqlStatementNoLog("START TRANSACTION");
````

File: interface/weno/formularyInsert.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    $ndc = $in[2];
+    $price = $in[5];
+    $drugName = $in[8];
+
+    sqlStatement("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
+    echo xlt("Inserted") . " " . text($drugName) . "<br> ";
+
+    $i++;
+} while ($i < $lines);
```
Comment (bradymiller) [inline interface/weno/formularyInsert.php]: To close out the above setting, then do the following her (after importing above data):
``````
// Settings to drastically speed up import with InnoDB
sqlStatementNoLog("COMMIT");
sqlStatementNoLog("SET autocommit=1");
``````

File: interface/weno/formularyInsert.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    $fileName = new SplFileObject($file);
+    $fileName->seek($i);
+ 
+    $in = explode("|", $fileName);
+
+    $ndc = $in[2];
+    $price = $in[5];
+    $drugName = $in[8];
+
+    sqlStatement("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
```
Comment (bradymiller) [inline interface/weno/formularyInsert.php]: Since do not need to log all these inserts, change to sqlStatementNoLog() function:
``````
sqlStatementNoLog("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));

File: interface/weno/formularyUpdate.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ 
+    if (!empty($find['ndc'])) {
+        sqlStatement("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
+        echo xlt("Inserted")."<br> ";
+        //sqlStatement("UPDATE erx_drug_paid SET price_per_unit = ? WHERE ndc = ?", array($price, $ndc));
+        //echo "Updated!<br>";
+    } else {
+        sqlStatement("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
+        echo xlt("Inserted")."<br> ";
+    }
```
Comment (bradymiller) [inline interface/weno/formularyUpdate.php]: I am a bit confused on this code block from line 45-53.
There is no difference between these 2 different queries. Can this just all be changed to:
``````
sqlStatementNoLog("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
echo xlt("Inserted")."<br> ";
``````
If we can simply this, then can use the innodb strategy used above to drastically increase the speed of the update.

File: interface/weno/markTx.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+$date = date("Y-m-d");
+$script = filter_input(INPUT_GET, "rx");
+
+$e_script = explode("-", $script);
+
+if ($e_script[0] === "NewRx") {
+    //See if the value is set
+    $check = "SELECT ntx FROM prescriptions WHERE id = ?";
+    $getVal = sqlStatement($check, array($e_script[1]));
+    $val = sqlFetchArray($getVal);
```
Comment (bradymiller) [inline interface/weno/markTx.php]: Since you are just grabbing one item, wouldn't sqlQuery be better? (then don't need the sqlFetchArray step)

File: library/ajax/drug_autocomplete/search.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    $term    = filter_input(INPUT_GET, "term");
+    
+    try {
+        $sql = "SELECT drug_label_name, price_per_unit FROM erx_drug_paid WHERE drug_label_name LIKE ? ";
+        $val = array('%'.$term.'%');
+        $res = sqlstatement($sql, $val);
+        while ($row = sqlFetchArray($res)) {
+            $return_arr[] =  $row['drug_label_name'] . " - ". $row['price_per_unit'];
+        }
+    } catch (PDOException $e) {
```
Comment (bradymiller) [inline library/ajax/drug_autocomplete/search.php]: This seems odd since the database connector is not even PDO???

File: library/ajax/pharmacy_autocomplete/search.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    try {
+        $sql = "SELECT id, Store_name, address_line_1, city, state FROM erx_pharmacies WHERE Store_name LIKE ? AND city LIKE ? ";
+        $sql .= " AND address_line_1 LIKE ? ";
+        $stm = array('%'.$term.'%','%'.$city.'%','%'.$address.'%');
+        $res = sqlstatement($sql, $stm);
+        while ($row = sqlFetchArray($res)) {
+            $return_arr[] =  $row['id'] . " - ".$row['Store_name'] . " ".$row['address_line_1']." ". $row['city'] ." " . $row['state'];
+        }
+    } catch (PDOException $e) {
```
Comment (bradymiller) [inline library/ajax/pharmacy_autocomplete/search.php]: This seems odd since the database connector is not even PDO???

File: library/weno/src/AdminProperties.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  /*
+  * Import the narcotics into the table database from SQL file
+  *
+  */
+        $sqlNarc = file_get_contents('../../contrib/weno/narc.sql');
+
+        sqlInsert($sqlNarc);
+
+          return "Narcotic drugs imported<br>";
```
Comment (bradymiller) [inline library/weno/src/AdminProperties.php]: should use xlt() on the string here

File: library/weno/src/AdminProperties.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    public function addNarcotics()
+    {
+
+  /*
+  * Import the narcotics into the table database from SQL file
+  *
+  */
+        $sqlNarc = file_get_contents('../../contrib/weno/narc.sql');
+
+        sqlInsert($sqlNarc);
```
Comment (bradymiller) [inline library/weno/src/AdminProperties.php]: how long does this import take?
If it takes awhile, let me know, and I can show you how to drastically reduce the time.

File: library/ajax/pharmacy_autocomplete/search.php (hunk 7 above)
Comment (juggernautsei) [reply library/ajax/pharmacy_autocomplete/search.php]: It is a left over from the original code.

File: interface/weno/markTx.php (hunk 5 above)
Comment (juggernautsei) [reply interface/weno/markTx.php]: Ok will do.

File: library/ajax/drug_autocomplete/search.php (hunk 6 above)
Comment (juggernautsei) [reply library/ajax/drug_autocomplete/search.php]: left over from the original code that was PDO.

File: library/weno/src/AdminProperties.php (hunk 9 above)
Comment (juggernautsei) [reply library/weno/src/AdminProperties.php]: That statement was from the first round before switching to the SQL version. I don't know how long it takes now.

File: library/ajax/drug_autocomplete/search.php (hunk 6 above)
Comment (bradymiller) [reply library/ajax/drug_autocomplete/search.php]: then would completely remove this try/catch (note sqlStatement outputs errors already)
as an aside, does these even work since the S in sqlstatement is not capitalized??

File: library/ajax/pharmacy_autocomplete/search.php (hunk 7 above)
Comment (bradymiller) [reply library/ajax/pharmacy_autocomplete/search.php]: then would completely remove this try/catch (note sqlStatement outputs errors already)
as an aside, does these even work since the S in sqlstatement is not capitalized??

File: library/weno/src/AdminProperties.php (hunk 9 above)
Comment (bradymiller) [reply library/weno/src/AdminProperties.php]: Are you saying this function is no longer used? If so, then would remove it.

File: library/weno/src/TransmitData.php [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    
+    public function getDrugList($pid, $date)
+    {
+
+        $sql = "SELECT * FROM prescriptions WHERE patient_id = ? AND ntx = 1 AND txDate = ?";
+        $res = sqlStatement($sql, array($pid, $date));
+        return $res;
+    }
+
+    public function checkList($send)
```
Comment (bradymiller) [inline library/weno/src/TransmitData.php]: This function doesn't appear to be used anywhere and is odd (to start with a sqlFetchArray). If it's not used, then would remove it.

File: templates/prescription/general_edit.html [hunk 11]
```diff
@@ -256,10 +265,24 @@
 {$ENDING_JAVASCRIPT}
 </script>
 </form>
+<!-- Has to be loaded after the form -->
```
Comment (bradymiller) [inline templates/prescription/general_edit.html]: Why did you move these to the bottom? Rec moving them to the top where include the css stuff. This is likely why getting the following js errors when open the prescription Add gui:
screenshot from 2017-09-02 21-18-38

File: templates/prescription/general_edit.html [hunk 12]
```diff
@@ -256,10 +265,24 @@
...
+<script type="text/javascript" src="{php} echo $GLOBALS['assets_static_relative']; {/php}/jquery-ui-1-12-1/jquery-ui.js"></script>
+
 {literal}
 <!-- for the fancy jQuery stuff -->
 <script type="text/javascript">
 
+$(function() {
+    
+    //Drug form autocomplete to text box.
+    $(".wDrug").autocomplete({
```
Comment (bradymiller) [inline templates/prescription/general_edit.html]: If this function is specific to weno rx, then would only output this if he weno rx feature is on.

File: templates/prescription/general_edit.html [hunk 13]
```diff
@@ -149,90 +156,92 @@
...
 </tr>
 <tr>
-	<td COLSPAN="1" class="text" ALIGN="right" VALIGN="MIDDLE" >{xl t='Starting Date'}</td>
-	<td COLSPAN="2" ALIGN="LEFT" VALIGN="MIDDLE" >
-		{html_select_date start_year="-10" end_year="+5" time=$prescription->start_date prefix="start_date_"}
-	</td>
+  <td COLSPAN="1" class="text" ALIGN="right" VALIGN="MIDDLE" >{xl t='Starting Date'}</td>
+  <td COLSPAN="2" ALIGN="LEFT" VALIGN="MIDDLE" >
+    {html_select_date start_year="-10" end_year="+5" time=$prescription->start_date prefix="start_date_"}
+    <input type="hidden" name="start_date" value="{$prescription->start_date}">
```
Comment (bradymiller) [inline templates/prescription/general_edit.html]: Did you add the above line for a bug fix? This is a new line of code. Note it's easier to see these things when look at this code while ignoring whitespace which can do with following link(note the `?w=1` at end of link):
https://github.com/openemr/openemr/pull/1017/files?w=1

**Comment by bradymiller** [comment]
Hi @juggernautsei ,
Just finished code review and testing. This is getting very close. Let me know when it's ready for the next review (guessing it will be ready for the main codebase at that point).
-brady

File: templates/prescription/general_edit.html (hunk 13 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: No, I did not add that line of code.  I have not recently edited this file.

File: templates/prescription/general_edit.html (hunk 11 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: I have not move anything in this area.

File: templates/prescription/general_edit.html (hunk 12 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: Ok, will do.

File: library/weno/src/TransmitData.php (hunk 10 above)
Comment (juggernautsei) [reply library/weno/src/TransmitData.php]: Removed

File: library/weno/src/AdminProperties.php (hunk 9 above)
Comment (juggernautsei) [reply library/weno/src/AdminProperties.php]: No, I am saying the words "Be patient this can take a while" can be removed.

File: library/ajax/pharmacy_autocomplete/search.php (hunk 7 above)
Comment (juggernautsei) [reply library/ajax/pharmacy_autocomplete/search.php]: Will remove try catch blocks.

File: library/ajax/drug_autocomplete/search.php (hunk 6 above)
Comment (juggernautsei) [reply library/ajax/drug_autocomplete/search.php]: Done

File: templates/prescription/general_edit.html (hunk 11 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: AAAARRRRRRR!!! I keep fixing this and it keeps getting broken :-(

Comment (juggernautsei) [reply templates/prescription/general_edit.html]: jquery-ui.min.css
This file is missing.

Comment (juggernautsei) [reply templates/prescription/general_edit.html]: Moved to the top and it works still so leaving and removing the note.

File: templates/prescription/general_edit.html (hunk 12 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: I don't know how to do this. I have tried <?php if ($GLOBALS['weno_rx_enable']) ?> and
 
{php} if ($GLOBALS['weno_rx_enable']) { {/php}

{php} } {/php}

I think because of the {literal} function no matter what is typed in there will be output.

**Comment by juggernautsei** [comment]
New updates are ready for inspection.

File: interface/weno/formularyInsert.php (hunk 1 above)
Comment (bradymiller) [reply interface/weno/formularyInsert.php]: still need to address this here
(also note the note I'll repost below that should be done after below code snipper)
(this will drastically increases the speed of this large import)

File: interface/weno/formularyInsert.php [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+    $ndc = $in[2];
+    $price = $in[5];
+    $drugName = $in[8];
+
+    sqlStatementNoLog("INSERT INTO `erx_drug_paid` SET `drug_label_name` = ?, `NDC` = ?, `price_per_unit` = ? ", array($drugName,$ndc,$price));
+    echo xlt("Inserted") . " " . text($drugName) . "<br> ";
+
+    $i++;
+} while ($i < $lines);
```
Comment (bradymiller) [inline interface/weno/formularyInsert.php]: below above while code, do the following:
``````
// Settings to drastically speed up import with InnoDB
sqlStatementNoLog("COMMIT");
sqlStatementNoLog("SET autocommit=1");

File: library/weno/src/AdminProperties.php (hunk 9 above)
Comment (bradymiller) [reply library/weno/src/AdminProperties.php]: Use the sqlStatementNoLog() function here instead.

File: library/weno/src/AdminProperties.php [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+    public function addNarcotics()
+    {
+
+  /*
+  * Import the narcotics into the table database from SQL file
+  *
+  */
+        $sqlNarc = file_get_contents('../../contrib/weno/narc.sql');
+
```
Comment (bradymiller) [inline library/weno/src/AdminProperties.php]: Above the sqlStatementNoLog, do the following:
``````
// Settings to drastically speed up import with InnoDB
sqlStatementNoLog("SET autocommit=0");
sqlStatementNoLog("START TRANSACTION");
``````

File: library/weno/src/AdminProperties.php [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        sqlInsert($sqlNarc);
+
```
Comment (bradymiller) [inline library/weno/src/AdminProperties.php]: below the sqlStatementNoLog(), do the following:
``````
// Settings to drastically speed up import with InnoDB
sqlStatementNoLog("COMMIT");
sqlStatementNoLog("SET autocommit=1");
``````

File: templates/prescription/general_edit.html (hunk 13 above)
Comment (bradymiller) [reply templates/prescription/general_edit.html]: git is clearly showing the above line is not in main codebase, but is in your code. If you did not add that line, then rec removing it.
If you want to get a clearer idea of what code you added to this file, check out the diff which ignores whitespace changes:
https://github.com/openemr/openemr/pull/1017/files?w=1

File: interface/weno/formularyInsert.php (hunk 14 above)
Comment (juggernautsei) [reply interface/weno/formularyInsert.php]: I usually just do what you ask no questions and try to understand later. Sometimes it is obvious after I read the instructions and I add it to my skills. 
This one is over my head enough that I want to ask a question. Placing the statements after the while loop means they will only execute after the loop is finished. How is that going to speed up the insert process?

File: library/weno/src/AdminProperties.php (hunk 15 above)
Comment (juggernautsei) [reply library/weno/src/AdminProperties.php]: Done

File: library/weno/src/AdminProperties.php (hunk 16 above)
Comment (juggernautsei) [reply library/weno/src/AdminProperties.php]: Done

File: interface/weno/formularyInsert.php (hunk 14 above)
Comment (bradymiller) [reply interface/weno/formularyInsert.php]: In this case, we are wrapping the large insert with settings to increase the innodb performance. Note you also need to place the settings above the while loop in my other post above. For inserts, innodb touches the hard drive after every insert for log/robust issues. This can be very expensive when inserting hundreds of records(for example, the setup script in openemr could take up to 45 minutes because of 100,000+ translations), so we are basically turning this off temporarily before the insert, and then turning it back on after the insert.

File: interface/weno/formularyInsert.php (hunk 1 above)
Comment (bradymiller) [reply interface/weno/formularyInsert.php]: @juggernautsei ,
Still need to address this. see my post below for details.

Comment (juggernautsei) [reply interface/weno/formularyInsert.php]: Added

File: templates/prescription/general_edit.html (hunk 13 above)
Comment (juggernautsei) [reply templates/prescription/general_edit.html]: Here is my proof that I did not add the lines 159-162. In the image below, I went back to where I did the dragndrop mod which has been untouched since then. The window on the left shows the file date. Look at the top of the editor, you can see that I am in the dragdrop folder. In the editor are the same lines from April of this year. This is from the main code base. I always fetch before creating a new branch. That code has been there long before April and I did not add it.  Can you be clear for me, you want me to remove lines 159-162?
not_me

Comment (bradymiller) [reply templates/prescription/general_edit.html]: only 1 line at 162. don't worry about this; it's minor and i'll fix it up when get this into the codebase (have fingers crossed that it will get into the codebase tonight :) )

File: interface/weno/formularyInsert.php (hunk 14 above)
Comment (juggernautsei) [reply interface/weno/formularyInsert.php]: understood.

**Comment by bradymiller** [comment]
hi @juggernautsei ,

I just brought in your feature into the codebase. thank you for your awesome contribution! I made a couple minor changes.

Here's your original code:
https://github.com/openemr/openemr/commit/049d45a2f0b1cccb62c2476cf6aaa315cba79d80

Here are a couple PSR2 changes:
https://github.com/openemr/openemr/commit/cec956607581557a5a0710c10c298513e56919cb

And here were a couple of other changes/fixes. (Mostly needed to fix the autocomplete feature so it didn't break the original autocomplete from the drugs table. So, when not using weno, it will use original drugs table autocomplete, and when using weno, it will autocomplete from your formulary table):
https://github.com/openemr/openemr/commit/04b6d41e384167015c7e6aa54179796e1865f314

-brady
